File: python/day12.py
import collections
import sys
from typing import Callable, Counter, Dict, List, Set, Tuple

from python.util import run_solution


Cave = str
Path = List[Cave]
Passage = Tuple[Cave, Cave]
Layout = Dict[Cave, Set[Cave]]
CaveVisitStrategy = Callable[[Cave, Counter[Cave]], bool]


def parse_passage(line: str) -> Passage:
    """..."""

    passage = line.split("-")
    assert (
        len(passage) == 2
    ), f"Expected 2 tokens on line '{line}', found {len(passage)}"

    return passage


def parse_layout(input: str) -> Layout:
    """..."""

    layout: Layout = collections.defaultdict(set)

    for line in input.splitlines():
        cave1, cave2 = parse_passage(line)
        layout[cave1].add(cave2)
        layout[cave2].add(cave1)

    return layout


def is_start(cave: str) -> bool:
    """..."""

    return cave == "start"


def is_end(cave: str) -> bool:
    """..."""

    return cave == "end"


def is_terminal(cave: str) -> bool:
    """..."""

    return is_start(cave) or is_end(cave)


def is_big_cave(cave: str) -> bool:
    """..."""

    return not is_terminal(cave) and all(letter.isupper() for letter in cave)


def is_small_cave(cave: str) -> bool:
    """..."""

    return not is_terminal(cave) and all(letter.islower() for letter in cave)


def find_paths(
    layout: Layout,
    path_so_far: Path,
    curr_cave: Cave,
    *,
    visit_count: Counter[Cave],
    can_visit_cave: CaveVisitStrategy,
) -> None:
    """..."""

    all_paths: List[Path] = []

    if not can_visit_cave(curr_cave, visit_count):
        return all_paths

    visit_count[curr_cave] += 1
    path_so_far.append(curr_cave)

    if is_end(curr_cave):
        all_paths.append(list(path_so_far))
    else:
        for neighbour in layout[curr_cave]:
            all_paths += find_paths(
                layout,
                path_so_far=path_so_far,
                curr_cave=neighbour,
                visit_count=visit_count,
                can_visit_cave=can_visit_cave,
            )

    path_so_far.pop()
    visit_count[curr_cave] -= 1

    return all_paths


def part1(input: str) -> int:
    """..."""

    def can_visit_cave(cave: Cave, prev_visit_counts: Counter[Cave]) -> bool:
        """..."""

        if prev_visit_counts[cave] == 0:
            return True

        if is_start(cave) or is_small_cave(cave):
            return False

        return True

    layout = parse_layout(input)
    paths = find_paths(
        layout,
        path_so_far=[],
        curr_cave="start",
        visit_count=collections.Counter(),
        can_visit_cave=can_visit_cave,
    )
    return len(paths)


def part2(input: str) -> int:
    """..."""

    def can_visit_cave(cave: Cave, prev_visit_counts: Counter[Cave]) -> bool:
        """..."""

        if prev_visit_counts[cave] == 0:
            return True

        if is_start(cave):
            return False

        if is_small_cave(cave):
            if prev_visit_counts[cave] > 1:
                return False

            if any(
                visit_count > 1
                for existing_cave, visit_count in prev_visit_counts.items()
                if is_small_cave(existing_cave) and existing_cave != cave
            ):
                return False

        return True

    layout = parse_layout(input)
    paths = find_paths(
        layout,
        path_so_far=[],
        curr_cave="start",
        visit_count=collections.Counter(),
        can_visit_cave=can_visit_cave,
    )

    return len(paths)


if __name__ == "__main__":
    sys.exit(run_solution(part1=part1, part2=part2))

File: typescript/day12.ts
import { runSolution } from "./util";

type Cave = string;
type Path = Cave[];
type Passage = [Cave, Cave];
type Layout = {
    [cave: Cave]: Set<Cave>;
};
type Counter<T> = Map<T, number>;
type CaveVisitStrategy = (cave: Cave, visitCount: Counter<Cave>) => boolean;

function parsePassage(line: string): Passage {
    const passage = line.split("-");
    if (passage.length !== 2) {
        throw new Error(
            `Expected 2 tokens on line '${line}', found ${passage.length}`
        );
    }

    return passage as Passage;
}

function parseLayout(input: string): Layout {
    const layout: Layout = {};

    input.split("\n").forEach((line) => {
        const [cave1, cave2]: [string, string] = parsePassage(line);

        if (!(cave1 in layout)) {
            layout[cave1] = new Set();
        }
        layout[cave1].add(cave2);

        if (!(cave2 in layout)) {
            layout[cave2] = new Set();
        }
        layout[cave2].add(cave1);
    });

    return layout;
}

const isStart = (cave: Cave) => cave === "start";
const isEnd = (cave: Cave) => cave === "end";
const isTerminal = (cave: Cave) => isStart(cave) || isEnd(cave);
const isSmallCave = (cave: Cave) =>
    !isTerminal(cave) &&
    Array.from(cave).every((letter) => letter.toLowerCase() === letter);
const isBigCave = (cave: Cave) =>
    !isTerminal(cave) &&
    Array.from(cave).every((letter) => letter.toUpperCase() === letter);

function findPaths(
    layout: Layout,
    pathSoFar: Path,
    currCave: Cave,
    visitCount: Counter<Cave>,
    canVisitCave: CaveVisitStrategy
): Path[] {
    const allPaths: Path[] = [];

    if (!canVisitCave(currCave, visitCount)) {
        return allPaths;
    }

    if (!visitCount.has(currCave)) {
        visitCount.set(currCave, 0);
    }
    visitCount.set(currCave, visitCount.get(currCave)! + 1);
    pathSoFar.push(currCave);

    if (isEnd(currCave)) {
        allPaths.push([...pathSoFar]);
    } else {
        layout[currCave].forEach((neighbour) => {
            allPaths.push(
                ...findPaths(
                    layout,
                    pathSoFar,
                    neighbour,
                    visitCount,
                    canVisitCave
                )
            );
        });
    }

    pathSoFar.pop();
    visitCount.set(currCave, visitCount.get(currCave)! - 1);

    return allPaths;
}

function part1(input: string): number {
    const canVisitCave: CaveVisitStrategy = (cave, prevVisitCounts) => {
        if (!prevVisitCounts.has(cave) || prevVisitCounts.get(cave)! === 0) {
            return true;
        }

        if (isStart(cave) || isSmallCave(cave)) {
            return false;
        }

        return true;
    };

    const layout = parseLayout(input);
    const paths = findPaths(layout, [], "start", new Map(), canVisitCave);

    return paths.length;
}

function part2(input: string): number {
    const canVisitCave: CaveVisitStrategy = (cave, prevVisitCounts) => {
        if (!prevVisitCounts.has(cave) || prevVisitCounts.get(cave)! === 0) {
            return true;
        }

        if (isStart(cave)) {
            return false;
        }

        if (isSmallCave(cave)) {
            const prevCountForCave = prevVisitCounts.get(cave);
            if (prevCountForCave !== undefined && prevCountForCave > 1) {
                return false;
            }

            const visitedOtherSmallCaveMoreThanOnce = Array.from(
                prevVisitCounts.entries()
            ).filter(
                ([existingSmallCave, visitCount]) =>
                    isSmallCave(existingSmallCave) &&
                    existingSmallCave !== cave &&
                    visitCount > 1
            );

            if (visitedOtherSmallCaveMoreThanOnce.length > 0) {
                return false;
            }
        }

        return true;
    };

    const layout = parseLayout(input);
    const paths = findPaths(layout, [], "start", new Map(), canVisitCave);

    return paths.length;
}

process.exit(runSolution(part1, part2));
